File: server/src/main/java/battle/model/weapons/gun/Gun.java
package battle.model.weapons.gun;

import battle.BattleMgr;
import battle.MapLogic;
import battle.model.Character;
import battle.model.Coordinate;
import battle.model.weapons.Weapon;
import config.battle.CharacterConfig;
import config.battle.WeaponsConfig;
import util.battle.BattleUtils;

public class Gun extends Weapon {
    public int width;

    public Gun(Character character, Coordinate currentPos, MapLogic mapMgr, BattleMgr battleMgr, double cellSize) {
        super(character, currentPos, mapMgr, battleMgr, cellSize);

        this.numUse = WeaponsConfig.TIME_USE_GUN;
        this.width = WeaponsConfig.WIDTH_GUN;
    }

    public int getTimeUseWeapon() {
        return this.numUse;
    }

    public void attack() {
        this.createBullet(this.player, this.curDir);
        this.numUse -= 1;
        this.player.updateTimeUseWeapon();
    }

    public void createBullet(Character player, Coordinate dirBullet) {
        // Tinh pos dau sung
        Coordinate pos = new Coordinate(this.currentPos.getX(), this.currentPos.getY());
        Coordinate dir = new Coordinate(this.curDir.getX(), this.curDir.getY());
        dir = BattleUtils.getInstance().normalizeVector(dir);
        pos = new Coordinate(pos.getX() + dir.getX() * this.width / 2, pos.getY() + dir.getY() * this.width / 2);

        // Add bullet
        Bullet bullet = new Bullet(player, pos, dirBullet, this.mapGUI, this.cellSize);
        this.mapGUI.listBullet.add(bullet);
    }
}

File: client/src/modules/Battle/Logic/Weapons/Gun/Gun.js
var Gun = Weapon.extend({

    ctor: function(player, pos, mapMgr, mapGUI, cellSize) {
        this._super(player, pos, mapMgr, mapGUI, cellSize, "res/gun.png");

        this.numUse = WEAPONS.TIME_USE_GUN
        this.width = WEAPONS.WIDTH_GUN;
    },

    getTimeUseWeapon: function () {
        return this.numUse
    },

    attack: function () {
        this.createBullet(this.player, this.curDir)
        this.numUse -= 1
        this.player.updateTimeUseWeapon()
    },

    createBullet: function (player, dirBullet) {
        // Tinh pos dau sung
        let pos = new cc.p(this.currentPos.x, this.currentPos.y);
        let dir = new cc.p(this.curDir.x, this.curDir.y);
        dir = cc.pNormalize(dir)
        pos = cc.pAdd(pos, cc.pMult(dir, this.width / 2))

        // Add bullet
        let bullet = new Bullet(player, pos, dirBullet, this.mapGUI, this.cellSize)
        this.mapGUI.listBullet.push(bullet)
        this.mapGUI.addChild(bullet)
    },
});